coaster/auth.py

```diff
@@ -5,43 +5,114 @@
+provides a ``_load_user()`` method, it will be called when :obj:`current_auth`
+is invoked for the first time in a request. Login managers can call
+:func:`add_auth_attribute` to load the actor (typically the authenticated user)
+and any other relevant authentication attributes.
 
-Login managers can use :func:`add_auth_attribute` to have these details
-handled for them.
+For compatibility with Flask-Login, a user object loaded at
+``_request_ctx_stack.top.user`` will be recognised and made available via
+:obj:`current_auth`.
 """
 
 from __future__ import absolute_import
+import collections
 from werkzeug.local import LocalProxy
 from flask import _request_ctx_stack, current_app, has_request_context
 
+__all__ = ['add_auth_attribute', 'add_auth_anchor', 'request_has_auth', 'current_auth']
 
-def add_auth_attribute(attr, value):
+
+def add_auth_attribute(attr, value, actor=False):
     """
     Helper function for login managers. Adds authorization attributes
-    that will be made available via :obj:`current_auth`.
+    to :obj:`current_auth` for the duration of the request.
+
+    :param str attr: Name of the attribute
+    :param value: Value of the attribute
+    :param bool actor: Whether this attribute is an actor
+       (user or client app accessing own data)
+
+    If the attribute is an actor and :obj:`current_auth` does not currently
+    have an actor, the attribute is also made available as
+    ``current_auth.actor``, which in turn is used by
+    ``current_auth.is_authenticated``.
+
+    The attribute name ``user`` is special-cased:
+
+    1. ``user`` is always treated as an actor
+    2. ``user`` is also made available as ``_request_ctx_stack.top.user`` for
+       compatibility with Flask-Login
     """
-    # Special-case 'user' for compatibility with Flask-Login
+    if attr in ('actor', 'anchors', 'is_anonymous', 'is_authenticated'):
+        raise AttributeError("Attribute name %s is reserved by current_auth" % attr)
+
+    # Invoking current_auth will also create it on the local stack. We can
+    # then proceed to set attributes on it.
+    ca = current_auth._get_current_object()
+    # Since :class:`CurrentAuth` overrides ``__setattr__``, we need to use :class:`object`'s.
+    object.__setattr__(ca, attr, value)
+
     if attr == 'user':
+        # Special-case 'user' for compatibility with Flask-Login
         _request_ctx_stack.top.user = value
-    else:
-        if not hasattr(_request_ctx_stack.top, 'auth'):
-            _request_ctx_stack.top.auth = {}
-        _request_ctx_stack.top.auth[attr] = value
+        # A user is always an actor
+        actor = True
+
+    if actor and ca.actor is None and value is not None:
+        object.__setattr__(ca, 'actor', value)
+
+
+def add_auth_anchor(anchor):
+    """
+    Helper function for login managers and view handlers to add a new auth anchor.
+    This is a placeholder until anchors are properly specified.
+    """
+    current_auth.anchors._add(anchor)
```

Comment:
I'm a little unclear about why `_add` is used here (and in the tests below). Is `AuthAnchors` supposed to be read-only?

Reply:
The request will have an initialisation phrase. Before_request loads the user from cookies, while decorators load anchors (access keys in a URL). These have to be added to current_auth. Once the main view function is underway, current_auth is read-only.

To make this work, current_auth does not publish write interfaces. Writing happens through the helper functions.